File: basic_dsa_one_hundred_sixty_seven.java
/**
 * Basic DSA Problems - Binary Tree Level Order Traversal
 * Return values level by level (BFS)
 */
import java.util.*;

public class basic_dsa_one_hundred_sixty_seven {

    // Problem: Given root, return level-order traversal as list of lists (each level one list).

    public static class TreeNode {
        int val;
        TreeNode left, right;
        TreeNode(int val) { this.val = val; }
    }

    /**
     * BFS with queue; for each level, poll all current size nodes and add their vals, then add children.
     * Time O(n), Space O(n)
     */
    public static List<List<Integer>> levelOrder(TreeNode root) {
        List<List<Integer>> out = new ArrayList<>();
        if (root == null) return out;
        Queue<TreeNode> q = new LinkedList<>();
        q.offer(root);
        while (!q.isEmpty()) {
            int size = q.size();
            List<Integer> level = new ArrayList<>();
            for (int i = 0; i < size; i++) {
                TreeNode node = q.poll();
                level.add(node.val);
                if (node.left != null) q.offer(node.left);
                if (node.right != null) q.offer(node.right);
            }
            out.add(level);
        }
        return out;
    }

    public static void main(String[] args) {
        System.out.println("=".repeat(60));
        System.out.println("Problem: Binary Tree Level Order Traversal");
        System.out.println("=".repeat(60));

        TreeNode root = new TreeNode(3);
        root.left = new TreeNode(9);
        root.right = new TreeNode(20);
        root.right.left = new TreeNode(15);
        root.right.right = new TreeNode(7);

        System.out.println("Level order -> " + levelOrder(root));
    }
}

File: basic_dsa_one_hundred_sixty_seven.py
"""
Basic DSA - Add Binary
Given two binary strings a and b, return their sum as a binary string.
"""


def add_binary(a, b):
    """
    Add from the end with carry.
    Time Complexity: O(n + m)
    Space Complexity: O(n + m)
    """
    i, j = len(a) - 1, len(b) - 1
    carry = 0
    out = []
    while i >= 0 or j >= 0 or carry:
        total = carry
        if i >= 0:
            total += ord(a[i]) - ord("0")
            i -= 1
        if j >= 0:
            total += ord(b[j]) - ord("0")
            j -= 1
        out.append("1" if (total & 1) else "0")
        carry = 1 if total >= 2 else 0
    out.reverse()
    return "".join(out)


if __name__ == "__main__":
    print("=" * 60)
    print("Problem: Add Binary")
    print("=" * 60)

    print(add_binary("11", "1"))  # 100
    print(add_binary("1010", "1011"))  # 10101
